### PR TITLE
Feature/capi

### PR DESCRIPTION
**Pull request by dezed**
1st csal commits.

### CODE REVIEW

File: clients/cppsal/include/sal_node.h [hunk 1]
```diff
@@ -0,0 +1,658 @@
...
+            NodeInfo(const std::string _cls, const std::string _group, const uint64_t _version,
+                     const std::string _name = "")
+                    : cls(_cls)
+                    , group(_group)
+                    , version(_version)
+                    , name(_name){};
+
+            std::string cls;   /// todo: make them const
+            std::string group; /// permission
+            uint64_t version;  // API version ?
```
Comment (CnlPepper) [inline clients/cppsal/include/sal_node.h:48]: Each SAL data class is versioned. This is not the SAL API version. This is defined in the datamodel.

File: clients/cppsal/include/sal_node.h [hunk 2]
```diff
@@ -0,0 +1,658 @@
...
+    {
+        typedef enum
+        {
+            NODE_BRANCH,
+            NODE_LEAF
+        } NodeType;
+
+
+        // TODO: those global names should be removed
+        // static char JSON_CONTENT_REPORT[] = "report"; // should be named as "summary"
```
Comment (CnlPepper) [inline clients/cppsal/include/sal_node.h:23]: Reports are not the same as a summary.
list() returns a report object that reports on the contents of a node.
get() returns the object contained in a node. A branch or a data object. The data object can be either a "full" object or a "summary" object as all data classes have two representations - the full data and a version containing only meta data (for use in tools visualising the data tree).

File: clients/cppsal/include/sal_node.h [hunk 3]
```diff
@@ -0,0 +1,658 @@
...
+            static TimeInfo decode(std::string modified_ts)
+            {
+                TimeInfo i;
+                int timeZoneDifferential;
+                i.modified = Poco::DateTimeParser::parse(TimeInfo::ts_format(), modified_ts, timeZoneDifferential);
+                return i;
+            }
+        };
+
+        /// integer version, how about timestamp, why not a fixed width int
```
Comment (CnlPepper) [inline clients/cppsal/include/sal_node.h:147]: Never do this. System time is not reliable. The server persistence layer should track a master tree revision and guarantee a revision number issued to any transaction is unique and sequential.

File: doc/source/conf.py [hunk 4]
```diff
@@ -19,43 +19,39 @@
 # If extensions (or modules to document with autodoc) are in another directory,
 # add these directories to sys.path here. If the directory is relative to the
 # documentation root, use os.path.abspath to make it absolute, like shown here.
-sys.path.insert(0, os.path.abspath('../..'))
+sys.path.insert(0, os.path.abspath("../.."))
```
Comment (CnlPepper) [inline doc/source/conf.py:22]: Whoever did this deserves to die.

File: clients/cppsal/include/sal_node.h (hunk 1 above)
Comment (dezed) [reply clients/cppsal/include/sal_node.h:48]: I cant comment on cppsal. This pull request is for the csal interface.

File: clients/cppsal/include/sal_node.h (hunk 2 above)
Comment (dezed) [reply clients/cppsal/include/sal_node.h:23]: I cant comment on cppsal. This pull request is for the csal interface.

File: clients/cppsal/include/sal_node.h (hunk 3 above)
Comment (dezed) [reply clients/cppsal/include/sal_node.h:147]: I cant comment on cppsal. This pull request is for the csal interface.

File: doc/source/conf.py (hunk 4 above)
Comment (dezed) [reply doc/source/conf.py:22]: I cant comment on cppsal. This pull request is for the csal interface.

**Comment by CnlPepper** [comment]
This PR is targeting the wrong branch, hence the confusion. The UKAEA work on new c/cpp client is being performed on the feature/sal_cpp branch. The feature/capi branch was the original development branch and is being left as a reference.

This PR should be against the feature/sal_cpp which will subsequently merge into development to prepare a release.

**Comment by dezed-ukaea** [comment]
Oh well. Nevermind. Your comments about the cpp interface will still need answering.

File: clients/cppsal/include/sal_node.h (hunk 1 above)
Comment (qingfengxia) [reply clients/cppsal/include/sal_node.h:48]: so it is kind of API version, not data revision version.   I think we can fix it for one release for all data classes.

File: clients/cppsal/include/sal_node.h (hunk 2 above)
Comment (qingfengxia) [reply clients/cppsal/include/sal_node.h:23]: I think I got the difference this week, as I am doing unit test, using data from server.  For list() there is only one version of json, called report.

File: clients/cppsal/include/sal_node.h (hunk 3 above)
Comment (qingfengxia) [reply clients/cppsal/include/sal_node.h:147]: got it, just some random thought recorded there

File: doc/source/conf.py (hunk 4 above)
Comment (qingfengxia) [reply doc/source/conf.py:22]: I have not touched any python code yet.  Although some bugs are found

Comment (CnlPepper) [reply doc/source/conf.py:22]: I suspect it must have been a misconfigured auto reformat tool.

File: clients/cppsal/include/sal_node.h (hunk 2 above)
Comment (CnlPepper) [reply clients/cppsal/include/sal_node.h:23]: You shouldn't need to infer how the system works from the server data, it's described in the datamodel and the SAL server rest API documentation.

Comment (CnlPepper) [reply clients/cppsal/include/sal_node.h:23]: https://simple-access-layer.github.io/documentation/server.html describes all the returned data, broken down by request.

File: clients/cppsal/include/sal_node.h (hunk 1 above)
Comment (CnlPepper) [reply clients/cppsal/include/sal_node.h:48]: Its the version of the data class itself. Each data class has a version number so the data class structures can be updated if a new attribute or modification is required e.g. Signal version 2 adds a new attribute. This allows a client to discriminate which version it is handling when returned by the server.

Old data stored in the storage array does not get promoted to the latest version (it is already written) so the client must be able to handle revisions of the data representation.

SAL is built to sit on top of object stores. The nature of the objects are allowed to evolve with use. Versioning is built into each and every level of the system to allow structured evolution and the ability to track the full history of changes.

The SAL REST API has a version
Each data classes have a version.
Each server subcomponent has a version.
The data tree is versioned (revisions)

**Pull request closed without merging**